### ISSUE TITLE
High CPU utilisation despite GPU usage

### ISSUE THREAD

**Issue by GlennDeLoose01** [ultralytics/ultralytics#10245]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Before delving into the problem, let's outline the context.
We ran the model on a GPU-cluster, running Kubernetes and Containerd. Despite the effort of reducing CPU usage and maximising the GPU utilisation, we are have trouble maximising the GPU usage while CPU being our bottleneck.

The outcome is as follows: We ran 3 instances, utilising yoloV8 classification/tracking on 30s recordings. The GPU allocated memory is about 1GB per instance. The main problem is the high CPU utilisation of 80-100% on all cores. Limiting the usage of more instances. 

We stripped down the code as much as possible to eliminate any problems on our side. Still encountering the same issue. The stripped down code is displayed below:

```
if torch.cuda.is_available():
    device = 'cuda' 
else:
    device = 'cpu'

model = YOLO("yolov8n.pt").to(device)
VIDEO_PATH = "data/video.mp4"

cap = cv2.VideoCapture(VIDEO_PATH)

**SNIP**
**Getting some video characteristics to set desired frame**

while frame_count < total_frames:
        # Frame position is set using previous frame_count information.
        cap.set(cv2.CAP_PROP_POS_FRAMES, frame_count)
        # Read the frame from the video.
        success, frame = cap.read()
              if success:
                   results = model.track(frame, persist=True, verbose=False)
                   print(results)

**SNIP**
**Update frame count**
```

The GPU memory is being allocated, and seems to be in use (comparing power usage, and processing speed of the video), but the CPU utilisation stays unusually high.

We have tried solutions from other issues reported, but none seem to help this problem. I'm not going to go in much more detail here, to not lead you down a possible wrong path. If there are any questions or things I forgot, feel free to ask.

Thanks in advance

Glenn

**Comment by glenn-jocher** [comment]
Hey Glenn 👋,

Thanks for reaching out and providing detailed insights into your setup and issue. High CPU utilization despite having GPU support can indeed be a challenge. Since you've already streamlined your code, a few additional things could be worth checking:

1. **Batch Size**: Increasing the batch size can sometimes help optimize GPU usage, though this depends on your available GPU memory.

2. **Data Loading and Preprocessing**: These operations are often run on the CPU. Employing a DataLoader with parameters like `num_workers` > 1 can help offload work from the main thread. PyTorch's DataLoader can be beneficial if you're looping through a dataset.

3. **Asynchronous Inference**: Using asynchronous calls for inference (if not already doing so) can improve overall throughput by not waiting for the GPU to finish processing before continuing with CPU-bound tasks.

4. **Monitoring Tools**: Utilize tools like `nvidia-smi` and PyTorch's `torch.cuda.profiler` to deeply analyze the GPU's utilization stats and identify potential bottlenecks.

5. **Version checks**: Ensure your PyTorch, CUDA, and cuDNN versions are all compatible and up-to-date.

If you're already batching or processing stream data, look into optimizing the stream loading and preprocessing phases. Frameworks like DALI from NVIDIA can dramatically reduce CPU load by offloading more data preprocessing to the GPU.

I hope these suggestions help you tackle the CPU bottleneck issue. If the problem persists or more specific guidance is needed, please don't hesitate to ask more questions. Good luck! 🚀

**Comment by GlennDeLoose01** [comment]
Hey Glenn,

Thank you for the quick response, we will definitely delve deeper into your suggestions.

We have discussed your answer internally, and we're wondering what we could expect from the CPU usage. Since the code example given is pretty basic, we would expect a very low CPU usage. Is it solely the preprocessing that requires this much CPU, is this within the expectations of the model? 

We have previously experimented with the yoloV3 model, and successfully limited the CPU usage to barely 1-3% running multiple instances. Without any of the above twists.

I have attached `htop` -screenshots and a `nvidia-smi` -screenshot to this comment. This is plain python, no k8s or containerised application.

We will have a look at PyTorch's `torch.cuda.profiler` and keep you updated. Once again thank you for the extended support.

<img width="644" alt="htop 1" src="https://github.com/ultralytics/ultralytics/assets/162997547/56a0e566-28b4-4ace-9092-75fc382be1fb">
<img width="1218" alt="htop 2" src="https://github.com/ultralytics/ultralytics/assets/162997547/38995c11-e9f6-4969-9daa-665916ac998e">
<img width="882" alt="Nvidia-smi" src="https://github.com/ultralytics/ultralytics/assets/162997547/d492c2bf-c1dc-427d-95a7-215b924cab06">

**Comment by glenn-jocher** [comment]
Hey there, Glenn! 🌟

Glad to hear you're diving into those suggestions! Regarding CPU usage, even basic code with YOLOv8 can lead to higher CPU utilization due to preprocessing steps and data loading, especially if they're not adequately optimized. YOLOv8, being more advanced and perhaps heavier than YOLOv3, might inherently demand more from the CPU for these operations. However, seeing CPU usage spike to 80-100% does seem a bit high for such tasks.

Comparing to YOLOv3, the architectural differences and the enhancements in YOLOv8 for accuracy and speed optimization could also mean different usage patterns of CPU and GPU resources. 

Since you mentioned moving from containerized applications to plain Python and still observing high CPU use, it suggests that the issue might not be related to how the environment is set up but rather how the model and data pipeline are interacting with your system's resources. 

As you plan to delve into `torch.cuda.profiler`, it should offer more insights into where the bottleneck lies. I highly recommend also reviewing data loading operations to see if there’s room for making them more efficient. Sometimes, employing techniques like batch processing (if not already) and ensuring data is preprocessed in a way that leverages GPU (where possible) can significantly reduce CPU load. 

Thanks for attaching the screenshots and for the continuous updates. They really help in understanding the situation better! Looking forward to hearing about your findings. Keep up the great work! 🚀

**Comment by GlennDeLoose01** [comment]
First and probably the last update:

- We updated the CUDA version from 11.4 to 12.4, this brought some new complications with docker and Kubernetes. This did solve the CPU usage problem partially.
- Another issue was from our side setting the CV2 capture, this apparently is less performant than reading and grabbing. Thanks to the `torch.cuda.profiler` and `cProfile` this could be seen easily.

I attached 2 new screenshots of the final performance, running 5 instances via Kubernetes of the yoloV8 model.

<img width="914" alt="Screenshot 2024-05-02" src="https://github.com/ultralytics/ultralytics/assets/162997547/799028ef-45c0-4b49-8164-624857e3583b">
<img width="793" alt="Screenshot" src="https://github.com/ultralytics/ultralytics/assets/162997547/6edc90f1-f2bc-4874-af28-8efdacb09b25">

With this I'll close this issue, thanks for the support!

**Comment by glenn-jocher** [comment]
Hey there! 🌈

Awesome news on the improvements! Upgrading CUDA and tweaking your CV2 capture method seem like solid steps towards enhancing performance. It's fantastic that `torch.cuda.profiler` and `cProfile` helped pinpoint those changes—just shows how powerful these tools can be. 🛠️

Thanks for attaching the final performance screenshots; it's great to see the practical effect of those updates, especially running 5 instances with improved efficiency. 🚀

If you encounter any more quirks or need assistance down the road, feel free to reach out. Closing the issue on such a positive note is fantastic, and we're here if you ever need more support. Thanks for sharing your journey with us!

All the best with your YOLOv8 adventures! 🌟

**Comment by ingGhost** [comment]
> First and probably the last update:
> 
> * We updated the CUDA version from 11.4 to 12.4, this brought some new complications with docker and Kubernetes. This did solve the CPU usage problem partially.
> * Another issue was from our side setting the CV2 capture, this apparently is less performant than reading and grabbing. Thanks to the `torch.cuda.profiler` and `cProfile` this could be seen easily.
> 
> I attached 2 new screenshots of the final performance, running 5 instances via Kubernetes of the yoloV8 model.
> 
> <img alt="Screenshot 2024-05-02" width="914" src="https://private-user-images.githubusercontent.com/162997547/327363528-799028ef-45c0-4b49-8164-624857e3583b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY3NzY1MjMsIm5iZiI6MTcyNjc3NjIyMywicGF0aCI6Ii8xNjI5OTc1NDcvMzI3MzYzNTI4LTc5OTAyOGVmLTQ1YzAtNGI0OS04MTY0LTYyNDg1N2UzNTgzYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxOVQyMDAzNDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00Y2VmMTdkYmExMmFkYzAzMTA1NWFhMjU5Yjk1Mzg2MjViYjQ2MTdiYTc3YTMxMTVhMjMzY2EzZmQ0OTFlNDBiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bsivSwr-lfAFsrRg0xngaf-8xUwJ3Kdu3qRUjEWpw-E"> <img alt="Screenshot" width="793" src="https://private-user-images.githubusercontent.com/162997547/327363514-6edc90f1-f2bc-4874-af28-8efdacb09b25.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY3NzY1MjMsIm5iZiI6MTcyNjc3NjIyMywicGF0aCI6Ii8xNjI5OTc1NDcvMzI3MzYzNTE0LTZlZGM5MGYxLWYyYmMtNDg3NC1hZjI4LThlZmRhY2IwOWIyNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxOVQyMDAzNDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00N2UwY2RjYzhmNWQyODUzMmNhM2ZlNTJiNTg3NDY5MThjM2JlNjBiZmM5OTdkZGM3MGMwNTkwODRhZmI2NjUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7ZS7YS1yZVpCvssHWlM8sOYtrNDNYT6zouuyjdO-rVk">
> With this I'll close this issue, thanks for the support!

hi, can u share code with improvement on CPU effect?

**Comment by glenn-jocher** [comment]
I'm glad to hear about your progress! Unfortunately, I can't share specific code improvements. However, focusing on optimizing data loading and preprocessing, as well as ensuring efficient use of CUDA, can significantly enhance performance. If you have further questions, feel free to ask!
